src/PostgREST/Query.hs

```diff
@@ -0,0 +1,234 @@
+  , update
+  , DbHandler
+  , ReadQueryResult
+  , WriteQueryResult(..)
+  ) where
+
+import Control.Monad.Except (liftEither)
+
+import qualified Hasql.DynamicStatements.Snippet as SQL
+import qualified Hasql.Transaction               as SQL
+import qualified Hasql.Transaction.Sessions      as SQL
+import qualified Network.HTTP.Types.Status       as HTTP
+
+import qualified PostgREST.DbStructure              as DbStructure
+import qualified PostgREST.Error                    as Error
+import qualified PostgREST.Query.QueryBuilder       as QueryBuilder
+import qualified PostgREST.Query.Statements         as Statements
+import qualified PostgREST.Request.ApiRequest       as ApiRequest
+import qualified PostgREST.Request.DbRequestBuilder as ReqBuilder
+
+import PostgREST.Config                  (AppConfig (..))
+import PostgREST.ContentType             (ContentType (..))
+import PostgREST.DbStructure             (DbStructure (..),
+                                          tablePKCols)
+import PostgREST.DbStructure.Identifiers (FieldName,
+                                          QualifiedIdentifier (..),
+                                          Schema)
+import PostgREST.DbStructure.Proc        (ProcDescription (..),
+                                          ProcVolatility (..))
+import PostgREST.DbStructure.Table       (Table)
+import PostgREST.Error                   (Error)
+import PostgREST.GucHeader               (GucHeader)
+import PostgREST.Request.ApiRequest      (Action (..),
+                                          ApiRequest (..),
+                                          InvokeMethod (..),
+                                          Target (..))
+import PostgREST.Request.Preferences     (PreferCount (..),
+                                          PreferParameters (..))
+import PostgREST.Request.Types           (MutateRequest, ReadRequest)
+
+import qualified PostgREST.DbStructure.Proc as Proc
+
+import Protolude hiding (Handler)
+
+type DbHandler = ExceptT Error SQL.Transaction
+
+type ReadQueryResult = (Statements.ResultsWithCount, Maybe Int64, [GucHeader], Maybe HTTP.Status)
+
+read :: ReadRequest -> Maybe FieldName -> AppConfig -> DbStructure -> ApiRequest -> DbHandler ReadQueryResult
+read req bField conf@AppConfig{..} dbStructure apiRequest@ApiRequest{..} = do
+  result@(tableTotal, queryTotal, _ , _, gucHeaders, gucStatus) <-
+    lift . SQL.statement mempty $
+      Statements.createReadStatement
+        (QueryBuilder.readRequestToQuery req)
+        (if iPreferCount == Just EstimatedCount then
+           -- LIMIT maxRows + 1 so we can determine below that maxRows was surpassed
+           QueryBuilder.limitedQuery countQuery ((+ 1) <$> configDbMaxRows)
+         else
+           countQuery
+        )
+        (iAcceptContentType == CTSingularJSON)
+        (shouldCount iPreferCount)
+        (iAcceptContentType == CTTextCSV)
+        bField
+        (pgVersion dbStructure)
+        configDbPreparedStatements
```

Comment:
```suggestion
    lift . SQL.statement mempty $
      Statements.createReadStatement
        (QueryBuilder.readRequestToQuery req)
        (if iPreferCount == Just EstimatedCount then
           -- LIMIT maxRows + 1 so we can determine below that maxRows was surpassed
           QueryBuilder.limitedQuery countQuery ((+ 1) <$> configDbMaxRows)
         else
           countQuery
        )
        (iAcceptContentType == CTSingularJSON)
        (shouldCount iPreferCount)
        (iAcceptContentType == CTTextCSV)
        bField
        (pgVersion dbStructure)
      configDbPreparedStatements
```

The prepared statements option seems to be an argument to `SQL.statement`, not `Statements.createReadStatement`. Might come up in other places, too?